opennms-config/src/main/java/org/opennms/netmgt/config/DefaultEventConfDao.java

```diff
@@ -253,6 +254,54 @@ public Events getRootEvents() {
 		return m_events;
 	}
 
+	@Override
+	public synchronized void loadEventsFromDB(List<EventConfEvent> dbEvents) {
+
+		// Group events by source and sort by source fileOrder
+		Map<String, List<EventConfEvent>> eventsBySource = dbEvents.stream()
+				.collect(Collectors.groupingBy(
+						event -> event.getSource().getName(),
+						LinkedHashMap::new,
+						Collectors.toList()
+				));
+
+		// Sort sources by fileOrder
+		List<Map.Entry<String, List<EventConfEvent>>> sortedSources = eventsBySource.entrySet().stream()
+				.sorted(Map.Entry.comparingByValue((events1, events2) -> {
+					Integer order1 = events1.get(0).getSource().getFileOrder();
+					Integer order2 = events2.get(0).getSource().getFileOrder();
+					return Integer.compare(order1 != null ? order1 : 0, order2 != null ? order2 : 0);
+				}))
+				.toList();
+
+		Events rootEvents = new Events();
+
+		// Create Events objects per source and add to m_loadedEventFiles
+		for (Map.Entry<String, List<EventConfEvent>> sourceEntry : sortedSources) {
+			String sourceName = sourceEntry.getKey();
+			List<EventConfEvent> sourceEvents = sourceEntry.getValue();
+
+			Events eventsForSource = new Events();
+
+			for (EventConfEvent dbEvent : sourceEvents) {
+				String xmlContent = dbEvent.getXmlContent();
+				if (xmlContent != null && !xmlContent.trim().isEmpty()) {
+					try {
+						Event event = JaxbUtils.unmarshal(Event.class, xmlContent);
```

Comment:
And call the the helper methods here
```suggestion
	 // Build Events per source
    for (Map.Entry<Source, List<EventConfEvent>> sourceEntry : sortedSources) {
        Events eventsForSource = buildEventsForSource(sourceEntry.getValue());
        rootEvents.addLoadedEventFile(sourceEntry.getKey().getName(), eventsForSource);
    }
```